File: csharp/HearthstoneReplays/Parser/ParserState.cs
#region

using System;
using System.Linq;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using HearthstoneReplays.Parser.ReplayData.Meta;
using HearthstoneReplays.Parser.ReplayData.Meta.Options;

#endregion

namespace HearthstoneReplays.Parser
{
	public class ParserState
	{
		public ParserState()
		{
			Reset();
		}

		public HearthstoneReplay Replay { get; set; }
		public Game CurrentGame { get; set; }
		public Node Node { get; set; }
		public GameData GameData { get; set; }
		public SendChoices SendChoices { get; set; }
		public Choices Choices { get; set; }
		public Options Options { get; set; }
		public Option CurrentOption { get; set; }
		public object LastOption { get; set; }
		public int FirstPlayerId { get; set; }
	    public int CurrentPlayerId { get; set; }
		public ChosenEntities CurrentChosenEntites { get; set; }

		public void Reset()
		{
			Replay = new HearthstoneReplay();
			CurrentGame = new Game();
		}

		public void UpdateCurrentNode(params Type[] types)
		{
			while(Node.Parent != null && types.All(x => x != Node.Type))
				Node = Node.Parent;
		}


	}
}

File: java/src/main/java/info/hearthsim/hsreplay/parser/ParserState.java
package info.hearthsim.hsreplay.parser;

import info.hearthsim.hsreplay.parser.replaydata.Game;
import info.hearthsim.hsreplay.parser.replaydata.GameData;
import info.hearthsim.hsreplay.parser.replaydata.HearthstoneReplay;
import info.hearthsim.hsreplay.parser.replaydata.gameactions.ChosenEntities;
import info.hearthsim.hsreplay.parser.replaydata.meta.Choices;
import info.hearthsim.hsreplay.parser.replaydata.meta.SendChoices;
import info.hearthsim.hsreplay.parser.replaydata.meta.options.Option;
import info.hearthsim.hsreplay.parser.replaydata.meta.options.Options;

import java.util.Arrays;

import lombok.Getter;
import lombok.Setter;
import lombok.ToString;
import lombok.extern.slf4j.Slf4j;

@Getter
@Setter
@ToString
@Slf4j
public class ParserState {

	private HearthstoneReplay replay;
	private Game currentGame;
	private Node node;
	private GameData gameData;
	private SendChoices sendChoices;
	private Choices choices;
	private Options options;
	private Option currentOption;
	private Object lastOption;
	private int firstPlayerId;
	private int currentPlayerId;
	private ChosenEntities currentChosenEntities;

	public ParserState() {
		reset();
	}

	public void reset() {
		replay = new HearthstoneReplay();
		currentGame = new Game();
	}

	public void updateCurrentNode(Class... types) {
		log.info("\t\tUpdating node " + node + " with types " + Arrays.toString(types));
		while (node.getParent() != null && Arrays.asList(types).stream().allMatch(x -> !node.getType().equals(x))) {
			node = node.getParent();
		}
	}

}
